### PR TITLE
{compiler}[GCCcore/9.3.0] AOCC v2.3.0

### PR DESCRIPTION
**Pull request by SebastianAchilles**
(created using `eb --new-pr`)

- [ ] Requires easybuilders/easybuild-easyblocks#2295

### CODE REVIEW

**Comment by SebastianAchilles** [comment]
WIP because `AOCC` picks the GCC from OS and not `GCCcore/9.3.0`. I think the `install.sh` needs a patch. You can check with:

```
$ ml AOCC
$ clang -v hello_world.c
...
Found candidate GCC installation: /usr/lib/gcc/x86_64-redhat-linux/8
Selected GCC installation: /usr/lib/gcc/x86_64-redhat-linux/8
...
```

Apart from that the easyconfig for `AOCC` is working.

**Comment by SebastianAchilles** [comment]
Test report by @sebastianachilles
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
ebvm.novalocal - Linux centos linux 8.3.2011, x86_64, Intel Xeon Processor (Skylake, IBRS), Python 3.6.8
See https://gist.github.com/af1741146c0c346555e440805de3cc85 for a full test report.

**Comment by boegel** [comment]
cc @robert-mijakovic

File: easybuild/easyconfigs/a/AOCC/AOCC-2.3.0-GCCcore-9.3.0.eb [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+sources = ['aocc-compiler-%(version)s.tar']
+checksums = [
+    '9f8a1544a5268a7fb8cd21ac4bdb3f8d1571949d1de5ca48e2d3309928fc3d15',  # aocc-compiler-2.3.0.tar
+]
+
+local_gccver = '9.3.0'
+
+dependencies = [
+    ('binutils', '2.34', '', ('GCCcore', local_gccver)),
```
Comment (boegel) [inline easybuild/easyconfigs/a/AOCC/AOCC-2.3.0-GCCcore-9.3.0.eb:21]: No need to hard specify compiler toolchain here, just use:

```python
dependencies = [
    ('binutils', '2.34'),
    ('ncurses', '6.2'),
    ('zlib', '1.2.11'),
    ('libxml2', '2.9.10'),
]

File: easybuild/easyconfigs/a/AOCC/AOCC-2.3.0-GCCcore-9.3.0.eb [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    ('ncurses', '6.2'),
+    ('zlib', '1.2.11'),
+    ('libxml2', '2.9.10', '', ('GCCcore', local_gccver)),
+]
+
+extract_sources = True
+
+install_cmd = 'bash install.sh && mkdir -p %(installdir)s && cp -r * %(installdir)s/'
```
Comment (boegel) [inline easybuild/easyconfigs/a/AOCC/AOCC-2.3.0-GCCcore-9.3.0.eb:29]: There's no clean install method? 😲 

If this is the only way, we should use `cp -a` to presersve timestamps, etc.

File: easybuild/easyconfigs/a/AOCC/AOCC-2.3.0-GCCcore-9.3.0.eb [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+version = '2.3.0'
+
+homepage = 'https://developer.amd.com/amd-aocc/'
+description = "AMD Optimized C/C++ & Fortran compilers (AOCC) based on LLVM 11.0"
+
+# Clang also depends on libstdc++ during runtime, but this dependency is
+# already specified as the toolchain.
+toolchain = {'name': 'GCCcore', 'version': '9.3.0'}
+
+sources = ['aocc-compiler-%(version)s.tar']
```
Comment (boegel) [inline easybuild/easyconfigs/a/AOCC/AOCC-2.3.0-GCCcore-9.3.0.eb:13]: No `source_urls` to enable auto-download?

**Review by boegel — CHANGES_REQUESTED** [review]

**Comment by boegel** [comment]
Maybe @migueldiascosta has some experience with this?

File: easybuild/easyconfigs/a/AOCC/AOCC-2.3.0-GCCcore-9.3.0.eb (hunk 1 above)
Comment (SebastianAchilles) [reply easybuild/easyconfigs/a/AOCC/AOCC-2.3.0-GCCcore-9.3.0.eb]: Thanks, I removed the hard specified compiler toolchain.

File: easybuild/easyconfigs/a/AOCC/AOCC-2.3.0-GCCcore-9.3.0.eb (hunk 2 above)
Comment (SebastianAchilles) [reply easybuild/easyconfigs/a/AOCC/AOCC-2.3.0-GCCcore-9.3.0.eb]: AOCC is a packed binary. `bash install.sh` is not installing the compiler, but creating a `setenv_AOCC.sh` that can be sourced. Maybe `bash install.sh` is not needed at all for easybuild. I have to check it again carefully. I followed closely the AOCC user guide to create a first easyconfig to have something to play around with it.

File: easybuild/easyconfigs/a/AOCC/AOCC-2.3.0-GCCcore-9.3.0.eb (hunk 3 above)
Comment (SebastianAchilles) [reply easybuild/easyconfigs/a/AOCC/AOCC-2.3.0-GCCcore-9.3.0.eb:13]: I was not able to get a working link. One needs to first accept the EULA and then can download the software.

**Comment by SebastianAchilles** [comment]
Test report by @sebastianachilles
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
ebvm.novalocal - Linux centos linux 8.3.2011, x86_64, Intel Xeon Processor (Skylake, IBRS), Python 3.6.8
See https://gist.github.com/9aabd05545680402ecb2f32e5d1ed2d8 for a full test report.

File: easybuild/easyconfigs/a/AOCC/AOCC-2.3.0-GCCcore-9.3.0.eb (hunk 3 above)
Comment (robert-mijakovic) [reply easybuild/easyconfigs/a/AOCC/AOCC-2.3.0-GCCcore-9.3.0.eb:13]: Actually, there is a direct link:
http://developer.amd.com/wordpress/media/files/aocc-compiler-2.3.0.tar

Comment (robert-mijakovic) [reply easybuild/easyconfigs/a/AOCC/AOCC-2.3.0-GCCcore-9.3.0.eb:13]: > I was not able to get a working link. One needs to first accept the EULA and then can download the software.

I experienced the same issue when I worked on the SPACK package for AOCC.
I decided to contacted AMD to tell me the 'real' location of the tarball.
They told me that they are going to do it themselves and let me know once it's done to review it.

Comment (robert-mijakovic) [reply easybuild/easyconfigs/a/AOCC/AOCC-2.3.0-GCCcore-9.3.0.eb:13]: I'm not sure about the legal aspect but EULA should be accepted somehow.
http://developer.amd.com/wordpress/media/files/AOCC_EULA.pdf

**Review by robert-mijakovic — DISMISSED** [review]

File: easybuild/easyconfigs/a/AOCC/AOCC-2.3.0-GCCcore-9.3.0.eb (hunk 2 above)
Comment (SebastianAchilles) [reply easybuild/easyconfigs/a/AOCC/AOCC-2.3.0-GCCcore-9.3.0.eb]: `bash install.sh` is just creating a `lua` module and a `setenv_AOCC.sh`. It is not needed for an installation with easybuild. I removed the `install_cmd`.

File: easybuild/easyconfigs/a/AOCC/AOCC-2.3.0-GCCcore-9.3.0.eb (hunk 3 above)
Comment (SebastianAchilles) [reply easybuild/easyconfigs/a/AOCC/AOCC-2.3.0-GCCcore-9.3.0.eb:14]: Thanks! I added the `source_urls`.

**Comment by SebastianAchilles** [comment]
> WIP because `AOCC` picks the GCC from OS and not `GCCcore/9.3.0`. I think the `install.sh` needs a patch. You can check with:
> 
> ```
> $ ml AOCC
> $ clang -v hello_world.c
> ...
> Found candidate GCC installation: /usr/lib/gcc/x86_64-redhat-linux/8
> Selected GCC installation: /usr/lib/gcc/x86_64-redhat-linux/8
> ...
> ```
> 

This seems to be not a bug! When installing `Clang` from source one can set `-DGCC_INSTALL_PREFIX=` to the software root of `GCCcore`. However AOCC is shipped as binary. That is why `Clang` is searching for GCC (actually it is `libstdc++` from `GCC`) in the standard location. The user can bypass this with the option --gcc-toolchain=$EBROOTGCCCORE`:

```
$ clang -v --gcc-toolchain=$EBROOTGCCCORE
...
Found candidate GCC installation: $EBROOTGCCCORE/lib/gcc/x86_64-pc-linux-gnu/9.3.0
Selected GCC installation: $EBROOTGCCCORE/lib/gcc/x86_64-pc-linux-gnu/9.3.0
```

**Comment by migueldiascosta** [comment]
> Maybe @migueldiascosta has some experience with this?

no, I hadn't looked into the compilers yet. One aspect I noticed just now is that this ships its own version of AMD LibM 

the recommended procedure for using a different version of AMD LibM seems to be to overwrite the one shipped with AOCC... (https://developer.amd.com/wp-content/resources/AOCC-2.3-Install-Guide.pdf)

so I guess we'll need to think about whether we want to tie the version of AMD LibM to the version of AOCC, and also what defaults to set regarding using glibc's libm or AMD's (?)




[1]

**Comment by SebastianAchilles** [comment]
@migueldiascosta The default is AMD libM. I tested this on Friday. The performance you get is quite good.

AOCC is shipped with the latest AMD LibM. But this could/might change in the future. An advantage of tying the version of AMD LibM to the version of AOCC is that it is more obvious which version is used. But I think this would require a custom easyblock, right?

**Comment by migueldiascosta** [comment]
@SebastianAchilles w.r.t. to libm performance, using a long loop of element-wise exponential of a vectors of 4 doubles, this is what I see (times in seconds, averages of 3 runs) on a Rome node:

  | -lm | -lamdlibm -lm
-- | -- | --
clang -O3 | 26 | 11.2
gcc -O3 | 11.9 | 11.7
icc -O3 | 5 | 
  |   |  
clang -O3 -ffast-math | 11.5 | 5.7 [a]
gcc -O3 -ffast-math | 2.3 [b] | 11.6

AOCC 2.3 (from this easyconfig), GCC 9.3.0, glibc 2.32 (from EESSI compatibility layer)

[a] `nm` does not show vector functions, so not sure if this is vectorized as documented
[b] `nm` does show vector functions, e.g. `_ZGVbN2v_exp`

as another reference, with `icc -O3 -march=core-avx2 -lm`, which automatically uses SVML, it runs in 3.5s.

**Comment by boegel** [comment]
Just so it's clear how to proceed with this: we should create a custom easyblock for AOCC which derives from `Binary`, calls the new `check_accepted_eula` method (see https://github.com/easybuilders/easybuild-framework/pull/3535), makes sure the "source" tarball is extract, specifies the custom environment variables to set, and  implements the (default) sanity check for AOCC (preferably including running something like `--help` on the important binaries).

**Comment by SebastianAchilles** [comment]
- [ ] Requires https://github.com/easybuilders/easybuild-easyblocks/pull/2295

To test easyconfig use `EB_ARGS=" --include-easyblocks-from-pr 2295 --accept-eula=AOCC"`

**Comment by SebastianAchilles** [comment]
Test report by @sebastianachilles
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/2295
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
ebvm.novalocal - Linux centos linux 8.3.2011, x86_64, Intel Xeon Processor (Skylake, IBRS), Python 3.6.8
See https://gist.github.com/5a2824631a6383607f84566bcbc85874 for a full test report.

**Comment by migueldiascosta** [comment]
Test report by @migueldiascosta
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/2295
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
sms - Linux centos linux 7.6.1810, x86_64, AMD EPYC 7601 32-Core Processor (zen), Python 2.7.5
See https://gist.github.com/14528178ab8dbd1c1990c66527e6a83a for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/2295
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
node3501.doduo.os - Linux RHEL 8.2, x86_64, AMD EPYC 7552 48-Core Processor (zen2), Python 3.6.8
See https://gist.github.com/f5baf1e86cc604362c847f165f8882b0 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/2295
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
node2710.swalot.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) CPU E5-2660 v3 @ 2.60GHz (haswell), Python 3.6.8
See https://gist.github.com/881dd3841fbd6ffb1482802953b12ccf for a full test report.

**Comment by boegel** [comment]
@boegelbot please test @ generoso
EB_ARGS="--include-easyblocks-from-pr 2295 --accept-eula=AOCC"

**Review by boegel — APPROVED** [review]

**Comment by boegel** [comment]
Going in, thanks @SebastianAchilles!

**Pull request merged**